### DISCUSSION TITLE
Sharp feature detection

### DISCUSSION THREAD

**Discussion by sloriot** [DGtal-team/DGtal#1633] [General] [answered]
Hi,

I was wondering whether there is something we can use to detect sharp features in a 3D segmented images?

Thanks!

cc @janetournois

**Comment by dcoeurjo** [comment]
Hi @sloriot,

How would you define sharp features in that context ? Union of sharp features ([_à la_ C1-discontinuities](https://perso.liris.cnrs.fr/david.coeurjolly/home/gallery/gallery/FullSizeRender_57.jpg)) for each interface between two labels?

**Comment by JacquesOlivierLachaud** [comment]
Hi
There are tools to compute curvature estimates (either digital VCM or corrected normal currents), so this is way to find sharp features on digital surfaces. Otherwise we worked (with Jeremy Levallois) on sharp feature detection using asymptotic linear geometry.
Or do you call sharp features places where three different objects meet ?

**Comment by janetournois** [comment]
Hello @dcoeurjo and @JacquesOlivierLachaud,
thanks for your quick replies!
The question was more about sharp features and C1-discontinuities, though places where more than 3 different objects meet would be interesting too.
Do you have a solution for that?
Jane.

**Comment by JacquesOlivierLachaud** [comment, marked as answer]
* For sharp features:
- Scale-space feature extraction on digital surfaces : https://www.sciencedirect.com/science/article/abs/pii/S0097849315000709
  (not sure we still have the implementation in DGtal, David ?)
* Curvature based sharp features estimation:
- digital Voronoi Covariance Measure: https://link.springer.com/chapter/10.1007/978-3-319-09955-2_12, 
    DGtal implementation https://dgtal.org/doc/stable/moduleVCM.html
   => not great for curvature, but eigenvalue analysis is quite stable for sharp  feature detection
- digital integral invariant: https://www.sciencedirect.com/science/article/abs/pii/S1077314214001003
   DGtal implementation  https://dgtal.org/doc/stable/moduleIntegralInvariant.html
   => ok for curvature, but features have to be thresholded and thinned.
- (interpolated) corrected normal current curvatures : https://hal.archives-ouvertes.fr/hal-02193774/ et https://onlinelibrary.wiley.com/doi/abs/10.1111/cgf.14067
  Pull Request en cours  dans DGtal https://github.com/DGtal-team/DGtal/pull/1617
  courbures plus précises, mais il reste à extraire les sharp features
* Discontinuity based sharp feature detection
   - piece-wise smooth regularization of normal vector fields with Ambrosio-Tortorelli functional
     => locate discontinuities of normal vector field, hence "sharp features".
     https://onlinelibrary.wiley.com/doi/abs/10.1111/cgf.13013
     DGtal implementation https://dgtal.org/doc/stable/moduleGenericAT.html

* Otherwise places with 3 objects met would be quite easy to detect in a labelled image. The problem is more to have a tracking of these places to define consistent contours, but is amenable with digital topology.

Hope this helps.
